### PR TITLE
split up io-util.[ch], and some clean-ups

### PR DESCRIPTION
**Pull request by poettering**
Some housekeeping, as preparation for #28891

### CODE REVIEW

File: src/basic/iovec-util.c [hunk 1]
```diff
@@ -14,6 +14,29 @@ size_t iovec_total_size(const struct iovec *i, size_t n) {
         return sum;
 }
 
+bool iovec_increment(struct iovec *i, size_t n, size_t k) {
+        /* Returns true if there is nothing else to send (bytes written cover all of the iovec),
+         * false if there's still work to do. */
+
```
Comment (bluca) [inline src/basic/iovec-util.c:22]: assert or check `i/n`

**Review by bluca — APPROVED** [review]

File: src/basic/iovec-util.c [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        assert(k == 0); /* Anything else would mean that we wrote more bytes than available,
+                         * or the kernel reported writing more bytes than sent. */
+        return true;
+}
+
+char* set_iovec_string_field(struct iovec *iovec, size_t *n_iovec, const char *field, const char *value) {
+        char *x;
+
```
Comment (YHNdnzj) [inline src/basic/iovec-util.c:44]: missing assertions?

File: src/basic/iovec-util.c [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        x = strjoin(field, value);
+        if (x)
+                iovec[(*n_iovec)++] = IOVEC_MAKE_STRING(x);
+        return x;
+}
+
+char* set_iovec_string_field_free(struct iovec *iovec, size_t *n_iovec, const char *field, char *value) {
+        char *x;
+
```
Comment (YHNdnzj) [inline src/basic/iovec-util.c:56]: also here

File: src/basic/iovec-wrapper.c [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        return mfree(iovw);
+}
+
+struct iovec_wrapper *iovw_free(struct iovec_wrapper *iovw) {
+        iovw_free_contents(iovw, false);
+
+        return mfree(iovw);
+}
+
+int iovw_put(struct iovec_wrapper *iovw, void *data, size_t len) {
```
Comment (YHNdnzj) [inline src/basic/iovec-wrapper.c:39]: here too

**Review by YHNdnzj — APPROVED** [review]
Assertions are missing in `iovec-wrapper.c`. If you touch this, can you please add them?

File: src/basic/iovec-wrapper.c (hunk 4 above)
Comment (YHNdnzj) [reply src/basic/iovec-wrapper.c:39]: (and below)

File: src/basic/iovec-util.c [hunk 5]
```diff
@@ -3,6 +3,16 @@
 #include "iovec-util.h"
 #include "string-util.h"
 
+size_t iovec_total_size(const struct iovec *i, size_t n) {
```
Comment (keszybz) [inline src/basic/iovec-util.c:6]: While at it, please rename `i` to `iovec`. (`i` really should be used only as an iterator.)

File: src/basic/iovec-util.h [hunk 6]
```diff
@@ -32,7 +32,8 @@ static inline bool IOVEC_INCREMENT(struct iovec *i, size_t n, size_t k) {
         return true;
 }
 
-#define IOVEC_NULL (struct iovec) {}
+#define IOVEC_NULL (const struct iovec) {}
```
Comment (keszybz) [inline src/basic/iovec-util.h:14]: Let's instead drop this. This is not really useful, in other similar cases we just do the cast in the open and the compiler can optimize the whole thing away.

File: src/basic/iovec-util.c [hunk 7]
```diff
@@ -14,6 +14,31 @@ size_t iovec_total_size(const struct iovec *i, size_t n) {
         return sum;
 }
 
+bool iovec_increment(struct iovec *i, size_t n, size_t k) {
```
Comment (keszybz) [inline src/basic/iovec-util.c:17]: Also here, `iovec` would be nicer as a name.

File: src/basic/iovec-util.h (hunk 6 above)
Comment (keszybz) [reply src/basic/iovec-util.h:14]: → https://github.com/systemd/systemd/pull/29679

**Pull request merged**